pkg/cloudprovider/providers/azure/azure_loadbalancer.go

```diff
@@ -1069,33 +1080,40 @@ func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service,
 	// update security rules: if the service uses a shared rule and is being deleted,
 	// then remove it from the shared rule
 	if useSharedSecurityRule(service) && !wantLb {
+		expectedSourceAddressPrefix, expectedSourceAddressPrefixes, err := az.getSourceAddressPrefix(sourceAddressPrefixes)
+		if err != nil {
+			return nil, err
+		}
 		for _, port := range ports {
-			for _, sourceAddressPrefix := range sourceAddressPrefixes {
-				sharedRuleName := az.getSecurityRuleName(service, port, sourceAddressPrefix)
-				sharedIndex, sharedRule, sharedRuleFound := findSecurityRuleByName(updatedRules, sharedRuleName)
-				if !sharedRuleFound {
-					klog.V(4).Infof("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
-					return nil, fmt.Errorf("Expected to find shared rule %s for service %s being deleted, but did not", sharedRuleName, service.Name)
-				}
-				if sharedRule.DestinationAddressPrefixes == nil {
-					klog.V(4).Infof("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
-					return nil, fmt.Errorf("Expected to have array of destinations in shared rule for service %s being deleted, but did not", service.Name)
-				}
-				existingPrefixes := *sharedRule.DestinationAddressPrefixes
-				addressIndex, found := findIndex(existingPrefixes, destinationIPAddress)
-				if !found {
-					klog.V(4).Infof("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
-					return nil, fmt.Errorf("Expected to find destination address %s in shared rule %s for service %s being deleted, but did not", destinationIPAddress, sharedRuleName, service.Name)
-				}
-				if len(existingPrefixes) == 1 {
-					updatedRules = append(updatedRules[:sharedIndex], updatedRules[sharedIndex+1:]...)
-				} else {
-					newDestinations := append(existingPrefixes[:addressIndex], existingPrefixes[addressIndex+1:]...)
-					sharedRule.DestinationAddressPrefixes = &newDestinations
-					updatedRules[sharedIndex] = sharedRule
-				}
-				dirtySg = true
+			sharedRuleName := ""
+			if len(expectedSourceAddressPrefixes) > 1 {
+				sharedRuleName = az.getSecurityRuleName(service, port, strings.Join(expectedSourceAddressPrefixes, "-"))
```

Comment:
`strings.Join(expectedSourceAddressPrefixes, "-")` would make the rule name different if sourceAddressPrefixes are updated (e.g. delete one by kubectl edit service). So the following `findSecurityRuleByName()` would always find nothing in this case?

Reply:
Yes unfortunately that's the behavior today. Since the rule name includes sourceAddressPrefixes, if you edit a service's source range, the cloud provider will think it's a new rule and create a new rule while the old rule with the old sourceAddressPrefixes in its name stays around forever.

With this PR, this behavior is only the case for shared rules. For rules that are not shared,  editing a service's sourceAddressPrefixes will not affect the rule name so it will still be deleted even if user's changed the service's sourceAddressPrefixes.

Should updating `load-balancer-source-ranges` as part of `kubectl edit service` be a supported scenario?

Reply:
yep, `loadBalancerSourceRanges` is part of service spec, it may be changed later after creation.

Reply:
> yep, loadBalancerSourceRanges is part of service spec, it may be changed later after creation.

But the existing behavior when loadBalancerSourceRanges is updated is that a new rule gets created and the old rule just lives on forever since the rule name can no longer be matched.

Reply:
oops, that should be a bug I didn't notice. We should fix that